### PR TITLE
Lock-free updates of the cached tombstone fragments

### PR DESCRIPTION
**Pull request by petermattis**
Replace the use of a `RWMutex` with atomic operations to update the
cached slice of tombstone fragments.

### CODE REVIEW

**Comment by petermattis** [comment]
This change is [Reviewable](https://reviewable.io/reviews/petermattis/pebble/57)

**Comment by petermattis** [comment]
Noticed this possibility in passing while looking at the RocksDB PR to support caching of fragmented range tombstones.

File: mem_table.go [hunk 1]
```diff
@@ -186,14 +185,23 @@ func (m *memTable) newRangeDelIter(*db.IterOptions) internalIterator {
...
-		// fragmented tombstones. If there was already a cached value, only
-		// overwrite it if we generated more tombstones.
-		if len(m.tombstones.vals) < len(tombstones) {
-			m.tombstones.vals = tombstones
+		for newPtr := unsafe.Pointer(&tombstones); true; {
+			oldPtr := atomic.LoadPointer(&m.tombstones.vals)
+			if oldPtr != nil {
+				// Adding tombstones to the memtable can only increase the number of
+				// fragmented tombstones. If there was already a cached value, only
+				// overwrite it if we generated more tombstones.
```
Comment (ajkr) [inline mem_table.go]: Nice, this is pretty convenient to resolve conflicts.

**Comment by ajkr** [comment]
What do you think about adding an pending invalidation atomic counter that works like this:

- The writer increments the value when an invalidation happens
- If a reader sees zero, then it can load from `m.tombstones.vals` and use those tombstones
- If a reader sees nonzero, it does fragmentation work creating its own `newPtr`. It then installs `newPtr` in the same way, CAS'ing it in only if it's longer than what's there currently. Then we decrement the atomic counter although not letting it go below zero.

**Comment by ajkr** [comment]
No, wait, I don't think that'll work because the pending invalidation counter may be decremented to zero by re-fragmentations that all began before the most recent invalidation happened.

**Comment by petermattis** [comment]
> What if there are two parallel writers, and a single reader that fragments tombstones before the second writer invalidates but updates the pointer after the second writer invalidates? Will think a bit more.

Yeah, I think something isn't quite right here.

**Comment by ajkr** [comment]
I think we could use load-link and store-conditional if it were available in Go. Something like:

- Writers store 0x0, same as now.
- A reader uses LL to read the pointer.
  - If it's non-0x0, just return it
  - Otherwise, fragment, then use SC to store the result. Whether it fails or succeeds, return the result of our fragmentation work. (If it fails that means either some other invalidation happened or another reader installed its fragmentation result. In the former case we can rely on a future reader to fragment again. In the latter case we know the pointer holds the fragmentation result for an invalidation at least as new as the one we're working on, so we don't need to do conflict checking.)

**Comment by petermattis** [comment]
@ajkr I'm not seeing how your proposal differs from the code in this PR. When the reader stores the fragmentation result, we need to be careful not to clear an invalidation that occurred after the reader produced the fragmentation.

I think we need some sort of invalidation sequence number. I think another level of indirection could accomplish this. `memTable.tombstones.vals` could be a `*tombstoneFrags`:

```
type tombstoneFrags struct {
  seqNum uint32 // count of range tombstones in the memTable
  vals unsafe.Pointer // *[]rangedel.Tombstones
}
```

Whenever a range tombstone is added, we replace atomically update `memtable.tombstones.vals` with a new `tombstoneFrags` if the previous frags struct contained a lower sequence number. In `memTable.newRangeDelIter` we retrieve the cached tombstones, populating `tombstoneFrags.vals` if it is `nil`.

In your scenario of 2 writers and 1 reader, the invalidation by the 2nd writer won't get lost by a reader constructing the fragments. There is an extra per-invalidation allocation in this approach, but that doesn't seem horrible. Another thought is to use `sync.Once` instead of atomically updating `tombstoneFrags.vals`.

**Comment by ajkr** [comment]
Yeah I think that should work. There should be no ABA problem between checking `seqNum` and CAS'ing `memtable.tombstones.vals`, right? For example, if during that gap another invalidation deleted the existing `tombstoneFrags` and created a new `tombstoneFrags` at the same memory address, then our CAS could mistakenly succeed. But I think we are holding a reference to the old `tombstoneFrags` throughout so its memory address shouldn't be reused.

**Comment by petermattis** [comment]
Changed the implementation per the description above, and added a test that fails with the previous implementation.

**Pull request merged**